controllers/networkchaos/trafficcontrol/types.go

```diff
@@ -188,73 +187,99 @@ func (r *Reconciler) Recover(ctx context.Context, req ctrl.Request, chaos v1alph
 func (r *Reconciler) cleanFinalizersAndRecover(ctx context.Context, networkchaos *v1alpha1.NetworkChaos) error {
 	var result error
 
+	source := networkchaos.Namespace + "/" + networkchaos.Name
+	m := podnetworkmanager.New(source, r.Log, r.Client)
+
 	for _, key := range networkchaos.Finalizers {
 		ns, name, err := cache.SplitMetaNamespaceKey(key)
 		if err != nil {
 			result = multierror.Append(result, err)
 			continue
 		}
 
-		var pod v1.Pod
-		err = r.Get(ctx, types.NamespacedName{
+		_ = m.WithInit(types.NamespacedName{
 			Namespace: ns,
 			Name:      name,
-		}, &pod)
-
-		if err != nil {
-			if !k8serror.IsNotFound(err) {
-				result = multierror.Append(result, err)
-				continue
-			}
-
-			r.Log.Info("Pod not found", "namespace", ns, "name", name)
-			networkchaos.Finalizers = utils.RemoveFromFinalizer(networkchaos.Finalizers, key)
-			continue
-		}
+		})
 
-		err = r.recoverPod(ctx, &pod, networkchaos)
+		err = m.Commit(ctx)
 		if err != nil {
-			result = multierror.Append(result, err)
-			continue
+			r.Log.Error(err, "fail to commit")
 		}
 
 		networkchaos.Finalizers = utils.RemoveFromFinalizer(networkchaos.Finalizers, key)
 	}
+	r.Log.Info("After recovering", "finalizers", networkchaos.Finalizers)
 
 	if networkchaos.Annotations[common.AnnotationCleanFinalizer] == common.AnnotationCleanFinalizerForced {
 		r.Log.Info("Force cleanup all finalizers", "chaos", networkchaos)
-		networkchaos.Finalizers = networkchaos.Finalizers[:0]
+		networkchaos.Finalizers = make([]string, 0)
 		return nil
 	}
 
 	return result
 }
 
-func (r *Reconciler) recoverPod(ctx context.Context, pod *v1.Pod, _ *v1alpha1.NetworkChaos) error {
-	r.Log.Info("Try to recover pod", "namespace", pod.Namespace, "name", pod.Name)
+func (r *Reconciler) applyTc(ctx context.Context, sources, targets []v1.Pod, externalTargets []string, m *podnetworkmanager.PodNetworkManager, networkchaos *v1alpha1.NetworkChaos) error {
+	for index := range sources {
+		pod := &sources[index]
 
-	pbClient, err := utils.NewChaosDaemonClient(ctx, r.Client, pod, common.ControllerCfg.ChaosDaemonPort)
-	if err != nil {
-		return err
+		key, err := cache.MetaNamespaceKeyFunc(pod)
+		if err != nil {
+			return err
+		}
+
+		networkchaos.Finalizers = utils.InsertFinalizer(networkchaos.Finalizers, key)
 	}
-	defer pbClient.Close()
 
-	if len(pod.Status.ContainerStatuses) == 0 {
-		return fmt.Errorf("%s %s can't get the state of container", pod.Namespace, pod.Name)
+	tcType := v1alpha1.Bandwidth
+	switch networkchaos.Spec.Action {
+	case v1alpha1.NetemAction, v1alpha1.DelayAction, v1alpha1.DuplicateAction, v1alpha1.CorruptAction, v1alpha1.LossAction:
+		tcType = v1alpha1.Netem
+	case v1alpha1.BandwidthAction:
+		tcType = v1alpha1.Bandwidth
+	default:
+		return fmt.Errorf("unknown action")
```

Comment:
how about save the `networkchaos.Spec.Action` in the error message

Reply:
Fixed